File: 2294-minimum-time-to-complete-trips/minimum-time-to-complete-trips.js
var minimumTime = function(time, totalTrips) {
    let i = 1
    let j = (Math.max(...time) * totalTrips)
    time.sort((a, b)=> a - b)

    function canTrip(time_taken) {
        let trips = 0
        for (let t of time) {
            trips += Math.floor(time_taken / t)
        }
        return trips >= totalTrips
    }

    while (i <= j) {
        let mid = Math.floor((i + j) / 2)
        if (canTrip(mid)) {
            j = mid - 1
        }
        else i = mid + 1
    }    
    return i
};

File: 2294-minimum-time-to-complete-trips/minimum-time-to-complete-trips.py
class Solution:
    def minimumTime(self, time: List[int], totalTrips: int) -> int:
        i = 1
        j = max(time) * totalTrips

        def canTrip(mid):
            completed_trips = 0
            for t in time:
                completed_trips += (mid // t)

            return completed_trips >= totalTrips
                
        while (i <= j):
            mid = (i + j) // 2

            if (canTrip(mid)):
                j = mid - 1
            else:
                i = mid + 1

        return i

        